otter/convergence/service.py

```diff
@@ -56,99 +62,363 @@ def all_met(server, current_lb_nodes):
+        eff = mark_divergent(tenant_id, group_id)
+        d = perform(self._dispatcher, eff)
+
+        def success(r):
+            log.msg('mark-dirty-success')
+            return r  # The result is ignored normally, but return it for tests
+        d.addCallbacks(success, log.err, errbackArgs=('mark-dirty-failure',))
+        return d
+
+
+class ConcurrentError(Exception):
+    """Tried to run an effect concurrently when it shouldn't be."""
+
+
+def make_lock_set():
+    """
+    Create a multi-lock function, which is a function that takes a key and an
+    effect, and runs the effect as long as no other multi-locked effect for the
+    same key is being run.
+
+    :return: a callable of (key, Effect) -> Effect, where the result of the
+        returned Effect will be the given effect's result, or an error of
+        :obj:`ConcurrentError` if the given key already has an effect being
+        performed by the same multi-lock function.
+    """
+    return partial(non_concurrently, Reference(pset()))
+
+
+@do
+def non_concurrently(locks, key, eff):
+    """
+    Run some Effect non-concurrently.
+
+    :param log: bound log
+    :param Reference locks: A reference to a PSet that will be used to record
+        which operations are currently being executed.
+    :param key: the key to use for this particular operation, which will be
+        stored in ``locks``
+    :param Effect eff: the effect to execute.
+
+    :return: Effect with the result of ``eff``, or an error of
+        :obj:`ConcurrentError` if the given key already has an associated
+        effect being performed.
+    """
+    if key in (yield locks.read()):
+        raise ConcurrentError(key)
+    yield locks.modify(lambda cc: cc.add(key))
+    try:
+        result = yield eff
+    finally:
+        yield locks.modify(lambda cc: cc.remove(key))
+    yield do_return(result)
+
+
+def get_my_divergent_groups(my_buckets, all_buckets):
+    """
+    Look up groups that are divergent and that are this node's
+    responsibility, according to ``my_buckets``.
+
+    :param my_buckets: collection of buckets allocated to this node
+    :param all_buckets: collection of all buckets
+
+    :returns: list of dicts, where each dict has ``tenant_id``,
+        ``group_id``, and ``version`` keys.
+    """
+    def structure_info(x):
+        # Names of the dirty flags are {tenant_id}_{group_id}.
+        path, stat = x
+        tenant, group = parse_dirty_flag(x[0])
+        return {'tenant_id': tenant, 'group_id': group,
+                'version': stat.version}
+
+    def got_children_with_stats(children_with_stats):
+        dirty_info = map(structure_info, children_with_stats)
+        num_buckets = len(all_buckets)
+        converging = (
+            info for info in dirty_info
+            if _stable_hash(info['tenant_id']) % num_buckets in my_buckets)
+        return list(converging)
+
+    eff = Effect(GetChildrenWithStats(CONVERGENCE_DIRTY_DIR))
```

Comment:
Non-blocking: I almost wanna say that you should get only children first and find the ones which this node _owns_ and then find their stats. But that may be an over-optimization not really required at this point. Its good to be aware though :)